src/nnet3/nnet-cudnn-simple-component.cc

```diff
@@ -0,0 +1,798 @@
+  backward_data_algo_(component.backward_data_algo_)
+ {
+  if (work_space_size_ != 0) {
+    work_space_ = CuDevice::Instantiate().Malloc(work_space_size_);
+  }
+}
+
+CuDNN3DConvolutionComponent::~CuDNN3DConvolutionComponent() {
+  CUDNN_SAFE_CALL(cudnnDestroyFilterDescriptor(filter_desc_));
+  CUDNN_SAFE_CALL(cudnnDestroyTensorDescriptor(bias_desc_));
+  CUDNN_SAFE_CALL(cudnnDestroyConvolutionDescriptor(conv_desc_));
+
+  if(work_space_ != NULL) {
+    CuDevice::Instantiate().Free(work_space_);
+  }
+}
+
+void CuDNN3DConvolutionComponent::Init(
+    int32 input_x_dim, int32 input_y_dim, int32 input_z_dim,
+    int32 filt_x_dim, int32 filt_y_dim, int32 filt_z_dim, int32 input_num_filters,
+    int32 filt_x_stride, int32 filt_y_stride, int32 filt_z_stride,
+    int32 num_filters,
+    int32 pad_x_dim, int32 pad_y_dim, int32 pad_z_dim,
+    int32 upscale_x_dim, int32 upscale_y_dim, int32 upscale_z_dim,
+    TensorVectorizationType input_vectorization,
+    BaseFloat param_stddev, BaseFloat bias_stddev) {
+  input_x_dim_ = input_x_dim;
+  input_y_dim_ = input_y_dim;
+  input_z_dim_ = input_z_dim;
+  input_num_filters_ = input_num_filters;
+  num_filters_ = num_filters;
+  int32 filters[kConvolutionDimension_ + 2];
+  filters[0] = num_filters_;
+  filters[1] = input_num_filters_;
+  filters[2] = filt_z_dim;
+  filters[3] = filt_y_dim;
+  filters[4] = filt_x_dim;
+  input_vectorization_ = input_vectorization;
+  int32 filter_dim = input_num_filters_ * filt_x_dim * filt_y_dim * filt_z_dim;
+  filter_params_.Resize(num_filters_, filter_dim, kUndefined,
+                        kStrideEqualNumCols);
+  bias_params_.Resize(num_filters_, kUndefined);
+  KALDI_ASSERT(param_stddev >= 0.0 && bias_stddev >= 0.0);
+  filter_params_.SetRandn();
+  filter_params_.Scale(param_stddev);
+  bias_params_.SetRandn();
+  bias_params_.Scale(bias_stddev);
+
+  int32 strides[kConvolutionDimension_];
+  strides[0] = filt_z_stride;
+  strides[1] = filt_y_stride;
+  strides[2] = filt_x_stride;
+
+  int32 upscales[kConvolutionDimension_];
+  upscales[0] = upscale_z_dim;
+  upscales[1] = upscale_y_dim;
+  upscales[2] = upscale_x_dim;
+
+  int32 padding[kConvolutionDimension_];
+  padding[0] = pad_z_dim;
+  padding[1] = pad_y_dim;
+  padding[2] = pad_x_dim;
+
+  CUDNN_SAFE_CALL(
+    cudnnSetConvolutionNdDescriptor(conv_desc_,
+                                    kConvolutionDimension_,
+                                    padding,
+                                    strides,
+                                    upscales,
+                                    CUDNN_CROSS_CORRELATION,
+                                    cudnn::GetDataType())
+  );
+
+  CUDNN_SAFE_CALL(cudnnCreateFilterDescriptor(&filter_desc_));
+  CUDNN_SAFE_CALL(
+    cudnnSetFilterNdDescriptor(filter_desc_,
+                               cudnn::GetDataType(),
+                               kConvolutionDimension_ + 2,
+                               filters
+                               )
+  );
```

Comment:
Here you are initializing a 5D filter descriptor but then you are setting/getting a 3D filter descriptor at other places, is there any problem here?